torchmultimodal/modules/layers/conv.py

```diff
@@ -0,0 +1,213 @@
+    stride does not divide into input evenly, then output = ceil(input / stride), following
+    the TensorFlow implementation explained here:
+    https://www.tensorflow.org/api_docs/python/tf/nn#notes_on_padding_2
+
+    Code taken from VideoGPT
+    https://github.com/wilson1yan/VideoGPT/blob/master/videogpt/vqvae.py
+
+    Args:
+        kernel_size (int or Tuple): size of convolutional kernel
+        stride (int or Tuple): stride amount of kernel
+        input_shape (Tuple or Size): tuple describing shape of input, without batch or channel dimension
+
+    Returns:
+        Tuple: the padding amount in a tuple of tuples for each dimension
+    """
+
+    convert_to_tuple = _ntuple(len(input_shape))
+    kernel_size = convert_to_tuple(kernel_size)
+    stride = convert_to_tuple(stride)
+
+    if not (len(kernel_size) == len(stride) == len(input_shape)):
+        raise ValueError("dims for kernel, stride, and input must match")
+
+    total_pad = []
+    for k, s, d in zip(kernel_size, stride, input_shape):
+        if d % s == 0:
+            pad = max(k - s, 0)
+        else:
+            pad = max(k - (d % s), 0)
+        total_pad.append(pad)
+    pad_input = []
+    for p in total_pad[::-1]:  # reverse since F.pad starts from last dim
+        pad_input.append(p // 2 + p % 2)
+        pad_input.append(p // 2)
+    pad_input = tuple(pad_input)
+    return pad_input
+
+
+def calculate_transpose_padding(
+    kernel_size: _size_any_t,
+    stride: _size_any_t,
+    input_shape: Union[Size, Tuple],
+    input_pad: _size_any_t = 0,
+) -> Tuple[Tuple, Tuple]:
+    """Calculates padding for transposed convolution based on input dims, kernel size, and stride.
+
+    Pads to match the 'SAME' padding in Keras, i.e., with a stride of 1 output is guaranteed
+    to have the same shape as input, with stride 2 the dimensions of output are doubled.
+
+    The 'padding' argument in ConvTranspose effectively trims the output, and the 'output_padding'
+    argument effectively expands the output. These two knobs are adjusted to meet desired output dim.
+
+    Args:
+        kernel_size (Tuple): size of convolutional kernel
+        stride (Tuple): stride amount of kernel
+        input_shape (Tuple or Size): tuple describing shape of input, without batch or channel dimension
+        input_pad (Tuple): amount of padding added to input, must be twice length of kernel/stride/input_shape
+
+    Returns:
+        Tuple: padding and output_padding to be used in ConvTranspose layers
+    """
+
+    convert_to_tuple = _ntuple(len(input_shape))
+    kernel_size = convert_to_tuple(kernel_size)
+    stride = convert_to_tuple(stride)
+
+    if not (len(kernel_size) == len(stride) == len(input_shape)):
+        raise ValueError("dims for kernel, stride, and input must match")
+    input_pad = _ntuple(len(input_shape) * 2)(input_pad)
+    if len(input_pad) % 2 != 0 or len(input_pad) // 2 != len(input_shape):
```

Comment:
it would fail if input_pad is already a tuple, which ntuple will not touch. then if the input_pad tuple is not the right length the assertion with fail. I will probably remove the use of PyTorch specific data types and remove ntuple, that way it's much clearer